src/prefect/agent/agent.py

```diff
@@ -125,20 +129,26 @@ def start(self) -> None:
             loop_intervals = {0: 0.25, 1: 0.5, 2: 1.0, 3: 2.0, 4: 4.0, 5: 8.0, 6: 10.0}
 
             index = 0
-            while not exit_event.wait(timeout=loop_intervals[index]):
-                self.heartbeat()
 
-                runs = self.agent_process(tenant_id)
-                if runs:
-                    index = 0
-                elif index < max(loop_intervals.keys()):
-                    index += 1
+            # the max workers default has changed in 3.5 and 3.8. For stable results the
+            # default 3.8 beahvior is elected here.
+            max_workers = min(32, (os.cpu_count() or 1) + 4)
 
-                self.logger.debug(
-                    "Next query for flow runs in {} seconds".format(
-                        loop_intervals[index]
+            with ThreadPoolExecutor(max_workers=max_workers) as executor:
```

Comment:
Potentially dumb question: in going with this implementation, is there a risk that a deadlock occurs if the agent finds `max_workers + 1` flow runs to submit?

Reply:
Not under that specific condition, no, there will always be a [future immediately returned by `submit`](https://github.com/python/cpython/blob/3.8/Lib/concurrent/futures/thread.py#L184-L189). There are two risks of deadlocks:
- during executor shutdown, if there is an underlying race with the [`_shutdown_lock`](https://github.com/python/cpython/blob/3.8/Lib/concurrent/futures/thread.py#L174)
- when putting work into the queue, if the [underlying file descriptors that make up the queue](https://github.com/python/cpython/blob/f84f65be5602e561fef04b66bb487fbc4e560db5/Lib/multiprocessing/connection.py#L527-L529) are in contention.
Though they would prevent `submit` from returning, both are unrelated to the value of max_workers and are very unlikely to occur.

I think the highest risk case is when the agent turns into a "black hole" for work: if the underlying workers never return, leaving new work to never be scheduled. The main thread would continue to pull work, backing up the queue further with work that wont be done. In this case, Lazarus will kick in and it will be scheduled again in which case the same work will not be queued again (due to the `self.submitting_flow_runs set`).